### DISCUSSION TITLE
How to get Failed step/scenario errors in variable at Environment file?

### DISCUSSION THREAD

**Discussion by jigisha9** [behave/behave#1140] [Q&A] [answered]
@jenisys @all 

How to get failed step/scenario errors in variable at Environment file?

I want to print errors/logs/exception message that we get during run time to be printed in Environment.py. How do i get errors for failed tests accessible from `environment.py` file.

I need this error message in after step/scenario so that I can send it to TestRail and update Testcase Run results. 

```python
# — FILE: features/environment.py
def after_step(context, step):
    pass  # Details left out here.

def after_scenario(context, scenario):
    pass  # Details left out here.
```

* Anybody have testrail/or any test mgt tool integration with Behave? 
* Or knows which inbuild function to use in after scenario that can return failed test error message?

FYI: I have used formatter and published HTML report having attachment/debug/runtime logs. I am not talking about this.

**Comment by jenisys** [comment, marked as answer]
@jigisha9 
As already stated:

* The natural solution for this is to use a special `Formatter`
* A `Formatter` is a extension-point in `behave`. Therefore, you can provide your own
* Multiple formatters can be used at the same time, but only one should use `stdout` as output destination. The other should write to files or send messages somewhere else or …

ALTERNATIVE:

* Use `JUnitXML` or `JSON` output and apply a post-processor that transforms the test-run information to your other test tool (test case management system, test database, …)
